File: cs/StrategyCore/StrategyEngine.cs
using System;
using System.Collections.Generic;
using System.Linq;
using MockTrader.Data;

namespace MockTrader.Core
{
    public sealed class StrategyConfig
    {
        public List<string> Factors { get; set; } = new List<string>(FactorKeys.All);
        public Dictionary<string, int> FactorSigns { get; set; } = new Dictionary<string, int> { ["momentum"] = 1, ["liquidity"] = 1, ["volume"] = 1, ["skewness"] = 1, ["rollYield"] = 1 };
        public string Combine { get; set; } = "equal";           // equal | ic | custom
        public Dictionary<string, double> FactorWeights { get; set; } // custom
        public int IcWindow { get; set; } = 60;
        public int IcHorizon { get; set; } = 5;
        public int LongCount { get; set; } = 5;
        public int ShortCount { get; set; } = 5;
        public string Mode { get; set; } = "longShort";
        public string Weighting { get; set; } = "equal";          // equal | score
        public bool Neutral { get; set; } = true;
        public string Rebalance { get; set; } = "monthly";
        public int RebalanceDays { get; set; } = 21;
        public int Buffer { get; set; } = 2;
        public double GrossExposure { get; set; } = 1.0;
        public int Warmup { get; set; } = 120;
    }

    public sealed class StrategyResult
    {
        public List<string> RebalanceDates { get; set; }
        public Dictionary<string, Dictionary<string, double>> Targets { get; set; } // date -> code -> weight
        public Dictionary<string, double?[]> Composite { get; set; }                // code -> array
        public StrategyConfig Config { get; set; }
    }

    public sealed class StrategyEngine
    {
        public StrategyResult Generate(FactorPanel panel, DataAccess ds, StrategyConfig cfg = null)
        {
            cfg = cfg ?? new StrategyConfig();
            var dates = panel.Dates;
            var varieties = panel.Varieties;
            var weights = FactorWeights(panel, ds, cfg);

            var composite = new Dictionary<string, double?[]>();
            foreach (var code in varieties) composite[code] = new double?[dates.Count];
            for (int t = 0; t < dates.Count; t++)
            {
                foreach (var code in varieties)
                {
                    double s = 0; int cnt = 0;
                    foreach (var f in cfg.Factors)
                    {
                        var zv = panel.Z[f][code][t];
                        if (zv == null) continue;
                        int sign = cfg.FactorSigns.TryGetValue(f, out var sg) ? sg : 1;
                        double w = weights.TryGetValue(f, out var wv) ? wv : 0;
                        s += sign * w * zv.Value; cnt++;
                    }
                    if (cnt > 0) composite[code][t] = s;
                }
            }

            int rebDays = cfg.Rebalance == "weekly" ? (cfg.RebalanceDays > 0 ? cfg.RebalanceDays : 5) : (cfg.RebalanceDays > 0 ? cfg.RebalanceDays : 21);
            var rebalanceDates = new List<string>();
            for (int t = cfg.Warmup; t < dates.Count; t += rebDays) rebalanceDates.Add(dates[t]);

            var targets = new Dictionary<string, Dictionary<string, double>>();
            var prevLong = new HashSet<string>();
            var prevShort = new HashSet<string>();

            foreach (var date in rebalanceDates)
            {
                int t = dates.IndexOf(date);
                var ranked = new List<(string code, double s)>();
                foreach (var code in varieties)
                {
                    var s = composite[code][t];
                    if (s != null && !double.IsNaN(s.Value) && !double.IsInfinity(s.Value)) ranked.Add((code, s.Value));
                }
                if (ranked.Count < Math.Max(cfg.LongCount, cfg.ShortCount) + 1) { targets[date] = new Dictionary<string, double>(); continue; }
                ranked.Sort((a, b) => b.s.CompareTo(a.s));

                var longNames = SelectSide(ranked, true, cfg, prevLong);
                var shortNames = SelectSide(ranked, false, cfg, prevShort);

                var wMap = new Dictionary<string, double>();
                double gross = cfg.GrossExposure;
                double sideGross = cfg.Neutral && cfg.Mode == "longShort" ? gross / 2.0 : gross;
                AssignWeights(wMap, longNames, +sideGross, cfg, composite, t);
                AssignWeights(wMap, shortNames, -sideGross, cfg, composite, t);
                targets[date] = wMap;
                prevLong = new HashSet<string>(longNames);
                prevShort = new HashSet<string>(shortNames);
            }
            return new StrategyResult { RebalanceDates = rebalanceDates, Targets = targets, Composite = composite, Config = cfg };
        }

        static List<string> SelectSide(List<(string code, double s)> ranked, bool isLong, StrategyConfig cfg, HashSet<string> prev)
        {
            int n = isLong ? cfg.LongCount : cfg.ShortCount;
            int buffer = cfg.Buffer;
            var pool = isLong
                ? ranked.Take(n + buffer).ToList()
                : ranked.Skip(Math.Max(0, ranked.Count - (n + buffer))).Reverse().ToList();
            var kept = pool.Where(x => prev.Contains(x.code)).Take(n).ToList();
            var keptSet = new HashSet<string>(kept.Select(x => x.code));
            var chosen = kept.Select(x => x.code).ToList();
            foreach (var x in pool)
            {
                if (chosen.Count >= n) break;
                if (!keptSet.Contains(x.code)) chosen.Add(x.code);
            }
            if (chosen.Count < n)
                foreach (var x in ranked) { if (chosen.Count >= n) break; if (!chosen.Contains(x.code)) chosen.Add(x.code); }
            return chosen.Take(n).ToList();
        }

        static void AssignWeights(Dictionary<string, double> weights, List<string> names, double sideGross, StrategyConfig cfg, Dictionary<string, double?[]> composite, int t)
        {
            if (names.Count == 0) return;
            if (cfg.Weighting == "score")
            {
                double s = 0;
                foreach (var code in names) s += Math.Abs(composite[code][t] ?? 0);
                if (s > 0)
                {
                    foreach (var code in names) weights[code] = sideGross * Math.Abs(composite[code][t] ?? 0) / s;
                    return;
                }
            }
            foreach (var code in names) weights[code] = sideGross / names.Count;
        }

        Dictionary<string, double> FactorWeights(FactorPanel panel, DataAccess ds, StrategyConfig cfg)
        {
            int n = cfg.Factors.Count;
            var outW = new Dictionary<string, double>();
            if (cfg.Combine == "custom" && cfg.FactorWeights != null)
            {
                double s = 0;
                foreach (var f in cfg.Factors) s += Math.Abs(cfg.FactorWeights.TryGetValue(f, out var v) ? v : 0);
                foreach (var f in cfg.Factors) outW[f] = s > 0 ? (cfg.FactorWeights.TryGetValue(f, out var v2) ? v2 : 0) / s : 1.0 / n;
                return outW;
            }
            if (cfg.Combine == "ic")
            {
                var ic = ComputeRollingIC(panel, ds, cfg);
                double s = 0;
                foreach (var f in cfg.Factors) s += Math.Abs(ic.TryGetValue(f, out var v) ? v : 0);
                if (s > 0) { foreach (var f in cfg.Factors) outW[f] = Math.Abs(ic.TryGetValue(f, out var v2) ? v2 : 0) / s; return outW; }
            }
            foreach (var f in cfg.Factors) outW[f] = 1.0 / n;
            return outW;
        }

        static Dictionary<string, double> ComputeRollingIC(FactorPanel panel, DataAccess ds, StrategyConfig cfg)
        {
            int horizon = cfg.IcHorizon, window = cfg.IcWindow;
            int T = panel.Dates.Count;
            var fwd = new Dictionary<string, double?[]>();
            foreach (var code in panel.Varieties)
            {
                var s = ds.GetSeries(code);
                fwd[code] = new double?[T];
                for (int t = 0; t + horizon < T; t++)
                    if (s.MainAdj[t] != null && s.MainAdj[t + horizon] != null && s.MainAdj[t] > 0)
                        fwd[code][t] = s.MainAdj[t + horizon].Value / s.MainAdj[t].Value - 1.0;
            }
            var ic = new Dictionary<string, double>();
            foreach (var f in cfg.Factors)
            {
                double best = 0;
                for (int t = T - 1; t >= window; t--)
                {
                    var xs = new List<double>(); var ys = new List<double>();
                    for (int tau = t - window + 1; tau <= t - horizon; tau++)
                        foreach (var code in panel.Varieties)
                        {
                            var zz = panel.Z[f][code][tau]; var rr = fwd[code][tau];
                            if (zz != null && rr != null && !double.IsNaN(zz.Value) && !double.IsNaN(rr.Value)) { xs.Add(zz.Value); ys.Add(rr.Value); }
                        }
                    if (xs.Count >= 10) { var rho = Stats.Spearman(xs, ys); if (!double.IsNaN(rho)) { best = Math.Abs(rho); break; } }
                }
                ic[f] = best > 0 ? best : 1.0 / cfg.Factors.Count;
            }
            return ic;
        }
    }
}

File: src/core/strategy/strategyEngine.js
/**
 * strategyEngine.js — 策略组合与信号生成 (S3: StrategyEngine "dll")。
 * 因子合成 -> 截面排序 -> 多空选品 -> 权重/中性化 -> 调仓(月度/周度 + 缓冲带)。
 * 全部参数可配置，默认等权、长 5 空 5、月度调仓、方向中性。
 */

import { spearman } from '../utils.js';

export const DEFAULT_STRATEGY_CONFIG = {
  factors: ['momentum', 'liquidity', 'volume', 'skewness', 'rollYield'],
  factorSigns: { momentum: 1, liquidity: 1, volume: 1, skewness: 1, rollYield: 1 },
  combine: 'equal', // 'equal' | 'ic' | 'custom'
  factorWeights: null, // 自定义权重 {factor: w}；combine='custom' 时生效
  icWindow: 60, // IC 加权：滚动窗口（交易日）
  icHorizon: 5, // IC 加权：前瞻收益天数（已做滞后，无未来函数）
  longCount: 5,
  shortCount: 5,
  mode: 'longShort', // 'longShort'(中性) | 'longOnly'
  weighting: 'equal', // 'equal' | 'score'
  neutral: true, // 方向中性（多空名义额相等）
  rebalance: 'monthly', // 'monthly' | 'weekly'
  rebalanceDays: 21, // 月度=21，周度=5
  buffer: 2, // 缓冲带（排名容忍）
  grossExposure: 1.0, // 总名义敞口 / 权益
  warmup: 120, // 起始调仓所需最小历史
};

/**
 * 计算合成得分面板：score[code][t] = Σ sign_f * w_f * z_f（缺失处为 null）。
 */
export function compositeScores(panel, config) {
  const { dates, varieties, z } = panel;
  const factors = config.factors || DEFAULT_STRATEGY_CONFIG.factors;
  const signs = Object.assign({}, config.factorSigns || {});
  const weights = factorWeights(panel, config);
  const score = {};
  for (const code of varieties) {
    score[code] = new Array(dates.length).fill(null);
  }
  for (let t = 0; t < dates.length; t++) {
    for (const code of varieties) {
      let s = 0;
      let cnt = 0;
      for (const f of factors) {
        const zv = z[f] && z[f][code] ? z[f][code][t] : null;
        if (zv == null) {
          continue;
        }
        const sign = signs[f] != null ? signs[f] : 1;
        const w = weights[f] != null ? weights[f] : 0;
        s += sign * w * zv;
        cnt++;
      }
      if (cnt > 0) {
        score[code][t] = s;
      }
    }
  }
  return score;
}

/** 因子权重：equal -> 1/n；custom -> 归一化；ic -> |IC| 归一化（滚动） */
export function factorWeights(panel, config) {
  const factors = config.factors || DEFAULT_STRATEGY_CONFIG.factors;
  const n = factors.length;
  const out = {};
  if (config.combine === 'custom' && config.factorWeights) {
    let s = 0;
    for (const f of factors) {
      s += Math.abs(config.factorWeights[f] || 0);
    }
    for (const f of factors) {
      out[f] = s > 0 ? (config.factorWeights[f] || 0) / s : 1 / n;
    }
    return out;
  }
  if (config.combine === 'ic' && panel._icWeights) {
    const ic = panel._icWeights;
    let s = 0;
    for (const f of factors) {
      s += Math.abs(ic[f] || 0);
    }
    if (s > 0) {
      for (const f of factors) {
        out[f] = Math.abs(ic[f] || 0) / s;
      }
      return out;
    }
  }
  for (const f of factors) {
    out[f] = 1 / n;
  }
  return out;
}

/** 滚动 IC（已滞后，无未来函数）：z_f(τ) vs 前瞻 horizon 收益，τ 取 [t-window, t-horizon] */
export function computeRollingIC(panel, ds, config) {
  const horizon = config.icHorizon || 5;
  const window = config.icWindow || 60;
  const { dates, varieties, z } = panel;
  const T = dates.length;
  // 前瞻收益 fwd[code][τ] = adjClose[τ+horizon]/adjClose[τ] - 1
  const fwd = {};
  for (const code of varieties) {
    const s = ds.getSeries(code);
    fwd[code] = new Array(T).fill(null);
    for (let t = 0; t + horizon < T; t++) {
      if (s.mainAdj[t] != null && s.mainAdj[t + horizon] != null && s.mainAdj[t] > 0) {
        fwd[code][t] = s.mainAdj[t + horizon] / s.mainAdj[t] - 1;
      }
    }
  }
  // 每个因子、每个调仓日 t：在 [t-window, t-horizon] 内与 fwd 的截面 spearman
  const factors = config.factors || DEFAULT_STRATEGY_CONFIG.factors;
  const icByFactor = {};
  for (const f of factors) {
    icByFactor[f] = new Array(T).fill(null);
  }
  for (let t = window; t < T; t++) {
    for (const f of factors) {
      const xs = [];
      const ys = [];
      for (let tau = t - window + 1; tau <= t - horizon; tau++) {
        for (const code of varieties) {
          const zz = z[f][code][tau];
          const rr = fwd[code][tau];
          if (zz != null && rr != null && Number.isFinite(zz) && Number.isFinite(rr)) {
            xs.push(zz);
            ys.push(rr);
          }
        }
      }
      if (xs.length >= 10) {
        const rho = spearman(xs, ys);
        icByFactor[f][t] = Number.isFinite(rho) ? rho : null;
      }
    }
  }
  return icByFactor;
}

export class StrategyEngine {
  /**
   * 生成信号：调仓日目标权重。
   * @returns {{rebalanceDates, targets, composite, config, icByFactor}}
   *   targets: { [date]: { [code]: weight(名义额/权益，正=多，负=空) } }
   */
  generate(panel, ds, config = {}) {
    const cfg = Object.assign({}, DEFAULT_STRATEGY_CONFIG, config);
    const dates = panel.dates;
    const varieties = panel.varieties;
    const factors = cfg.factors || DEFAULT_STRATEGY_CONFIG.factors;

    // IC 权重（可选）
    let icByFactor = null;
    if (cfg.combine === 'ic') {
      icByFactor = computeRollingIC(panel, ds, cfg);
      panel._icWeights = {};
      for (const f of factors) {
        const arr = icByFactor[f];
        // 取最近一个非空 IC 的 |IC| 作为权重
        let w = 0;
        for (let t = arr.length - 1; t >= 0; t--) {
          if (arr[t] != null) {
            w = Math.abs(arr[t]);
            break;
          }
        }
        panel._icWeights[f] = w || 1 / factors.length;
      }
    }

    const score = compositeScores(panel, cfg);

    // 调仓日序列（交易日计数）
    const rebDays = cfg.rebalance === 'weekly' ? cfg.rebalanceDays || 5 : cfg.rebalanceDays || 21;
    const rebalanceDates = [];
    for (let t = cfg.warmup; t < dates.length; t += rebDays) {
      rebalanceDates.push(dates[t]);
    }

    const targets = {};
    let prevLong = new Set();
    let prevShort = new Set();

    for (const date of rebalanceDates) {
      const t = dates.indexOf(date);
      const ranked = [];
      for (const code of varieties) {
        const s = score[code][t];
        if (s != null && Number.isFinite(s)) {
          ranked.push({ code, s });
        }
      }
      if (ranked.length < Math.max(cfg.longCount, cfg.shortCount) + 1) {
        targets[date] = {};
        continue;
      }
      ranked.sort((a, b) => b.s - a.s);

      const longNames = selectSide(ranked, 'long', cfg, prevLong);
      const shortNames = selectSide(ranked, 'short', cfg, prevShort);

      const weights = {};
      const gross = cfg.grossExposure || 1.0;
      const sideGross = cfg.neutral && cfg.mode === 'longShort' ? gross / 2 : gross;
      assignWeights(weights, longNames, +sideGross, cfg, score, t);
      assignWeights(weights, shortNames, -sideGross, cfg, score, t);

      targets[date] = weights;
      prevLong = new Set(longNames);
      prevShort = new Set(shortNames);
    }

    return { rebalanceDates, targets, composite: score, config: cfg, icByFactor };
  }
}

/** 缓冲带选品 */
function selectSide(ranked, side, cfg, prev) {
  const n = side === 'long' ? cfg.longCount : cfg.shortCount;
  const buffer = cfg.buffer || 0;
  const pool =
    side === 'long' ? ranked.slice(0, n + buffer) : ranked.slice(-(n + buffer)).reverse();
  const poolSet = new Set(pool.map((x) => x.code));
  const kept = pool.filter((x) => prev.has(x.code)).slice(0, n);
  const keptSet = new Set(kept.map((x) => x.code));
  const filled = [];
  for (const x of pool) {
    if (filled.length >= n - kept.length) {
      break;
    }
    if (!keptSet.has(x.code)) {
      filled.push(x);
    }
  }
  const chosen = kept.concat(filled).map((x) => x.code);
  // 若仍不足 n，从 pool 外补
  if (chosen.length < n) {
    const rest = side === 'long' ? ranked : ranked.slice().reverse();
    for (const x of rest) {
      if (chosen.length >= n) {
        break;
      }
      if (!chosen.includes(x.code) && poolSet.has(x.code)) {
        continue;
      }
      if (!chosen.includes(x.code)) {
        chosen.push(x.code);
      }
    }
  }
  return chosen.slice(0, n);
}

function assignWeights(weights, names, sideGross, cfg, score, t) {
  if (!names.length) {
    return;
  }
  if (cfg.weighting === 'score') {
    let s = 0;
    for (const code of names) {
      s += Math.abs(score[code][t] || 0);
    }
    if (s > 0) {
      for (const code of names) {
        weights[code] = (sideGross * Math.abs(score[code][t] || 0)) / s;
      }
      return;
    }
  }
  for (const code of names) {
    weights[code] = sideGross / names.length;
  }
}
